srk/src/numberField.ml

```diff
@@ -0,0 +1,994 @@
+  val add : elem -> elem -> elem
+  val sub : elem -> elem -> elem
+  val inverse : elem -> elem
+  val exp : elem -> int -> elem
+  val equal : elem -> elem -> bool
+  val compare : elem -> elem -> int
+  val is_zero : elem -> bool
+  val one : elem
+  val zero : elem
+  val negate : elem -> elem
+  val pp : Format.formatter -> elem -> unit
+  module X : 
+    sig
+      include Euclidean with type scalar = elem
+      val pp : Format.formatter -> t -> unit
+      val lift : QQX.t -> t
+      val de_lift : t -> QQXs.t
+      val factor_square_free_poly : t -> (scalar * ((t * int) list))
+      val factor : t -> (scalar * ((t * int) list))
+      val extract_root_from_linear : t -> elem
+    end
+  module O : sig
+    type ideal
+    type o
+    val pp_o : Format.formatter -> o -> unit
+    val make_o_el : elem -> ZZ.t * o
+    val order_el_to_f_elem : ZZ.t * o -> elem
+    val idealify : ZZ.t array array -> ideal
+    val pp_i : Format.formatter -> ideal -> unit
+    val equal_i : ideal -> ideal -> bool
+    val sum_i : ideal -> ideal -> ideal
+    val mul_i : ideal -> ideal -> ideal
+    val ideal_generated_by : o -> ideal
+    val one_i : ideal
+    val intersect_i : ideal -> ideal -> ideal
+    val quotient_i : ideal -> ideal -> ideal
+    val get_smallest_int : ideal -> ZZ.t
+    type frac_ideal
+    val one : frac_ideal
+    val sum : frac_ideal -> frac_ideal -> frac_ideal
+    val intersect : frac_ideal -> frac_ideal -> frac_ideal
+    val mul : frac_ideal -> frac_ideal -> frac_ideal
+    val exp : frac_ideal -> int -> frac_ideal
+    val quotient : frac_ideal -> frac_ideal -> frac_ideal
+    val pp : Format.formatter -> frac_ideal -> unit
+    val equal : frac_ideal -> frac_ideal -> bool
+    val subset : frac_ideal -> frac_ideal -> bool
+    val make_frac_ideal : ZZ.t -> ideal -> frac_ideal
+    val compute_overorder : frac_ideal -> frac_ideal -> frac_ideal * frac_ideal
+    val factor_refinement : frac_ideal list -> frac_ideal * (frac_ideal * int) list
+    val compute_factorization : frac_ideal list -> int list list * frac_ideal list * frac_ideal
+    val find_unit_basis : o list -> int list list
+  end
+  val find_relations : elem list -> int list list
+end
+
+
+
+
+
+
+module MakeNF (A : sig val min_poly : QQX.t end) = struct
+
+  type elem = QQX.t (* elements of the field are polynomials of degree < the degree of the field*)
+
+  let compute_min_poly_p el p = 
+    let open Linear in
+    let p_deg = QQX.order p in
+    if p_deg = 0 then el
+    else
+      let m = Array.make_matrix p_deg (p_deg+1) QQ.zero in (*m is a matrix whose columns consist of the powers of el*)
+      for i = 0 to p_deg do
+        let r = snd (QQX.qr (QQX.exp el i) p) in
+        BatEnum.iter (
+          fun (c, d) -> 
+            m.(d).(i) <- c
+        ) (QQX.enum r)
+      done;
+      (*In general I don't think null is guarenteed to be reduced with respect to powers of el. However,
+         I think it is do to how nullspace is implemented. *)
```

Comment:
Yes, that is the assumption. Take the example of Q[x]/p(x) where p(x) = x^4 - 2x^2 + 9. Then find the minimal polynomial of i = (x^3 + x)/6. i is really the imaginary unit. Taking powers of i and then computing the nullspace gives, the following possible basis. (1, 0, 1, 0, 0), (0, 1, 0, 1, 0), (-1, 0, 0, 0, 1). Corresponding to i^2 = -1, i^3 = -i, and i^4 = 1. The first basis element is minimal because the basis has been oriented to put higher powers to the right. If we use rref to compute the null space we can be sure that the minimal vector will be in the basis. nullspace uses rref so I think the assumption holds. However, this property is not stated for the interface of nullspace, so if the implementation of nullspace changes the assumption might not hold. I used nullspace because I couldn't find an accessible interface to compute rref.

Reply:
One way to solve this without this assumption:
`module M = Linear.MakeLinearMap(QQ)(SrkUtil.Int)(QQX)(QQX)`
creates a module M representing linear maps Q[x] -> Q[x].  You can construct a linear map by adding bindings p^0 -> 1, p^1 -> x, p^2 -> x^2 to a map m ... until eventually `M.apply m p^n` is defined (returns a polynomial rather than None).  The returned polynomial is the minimal polynomial.

Reply:
I don't think I understand. I tried this

let compute_min_poly_p el p = 
  let p_deg = QQX.order p in
  if p_deg = 0 then el
  else
    let e i = snd (QQX.qr (QQX.exp el i) p) in
    let rec aux app_res m i = 
      match app_res with
      | Some p -> p
      | None -> 
        let new_m = M.add_exn (e i) (QQX.exp (QQX.identity) i) m in
        aux (M.apply new_m (e p_deg)) new_m (i+1)
    in
    let min = aux None M.empty 0 in
    let lc = QQX.coeff (QQX.order min) min in
    QQX.map (fun _ c -> QQ.div c lc) min

But that doesn't seem to work. I'm thinking I'm misunderstanding M.apply and M.add

Reply:
What I had in mind was something like
```
let compute_min_poly_p el p =
    let e i = snd (QQX.qr (QQX.exp el i) p) in
    let rec aux app_res m i =
      match app_res with
     | Some p -> QQX.add_term QQ.one i p
     | None ->
       let new_m = M.add_exn (e i) (QQX.negate (QQX.exp (QQX.identity) i)) m in
         aux (M.apply new_m (e (i+1)) new_m (i+1)
    in
    aux None M.empty 0
```
The main difference being that we apply the map to `p^{i+1}` instead of `p^p_deg`.  The idea being that if p^{i+1} is in the domain of new_m. then p^{i+1} is a linear combination of p^0,...,p^i, and `M.apply new_m p^{i+1}` gives a polynomial q (of degree i) such that q(p) = -p^{i+1}.  The minimal polynomial is then x^{i+1} + q.

(I also don't understand the special-case reasoning for p_deg = 0.  I would expect the min polynomial of (say) 42 to be x-42 rather than 6, and I think the general case covers this).

Reply:
I've updated compute_min_poly_p to what you have. Is there an efficiency difference between the two implementations? I'm not sure how M.add and M.apply work. Before the time was taken by a single rref call.

As to the special case for p_deg = 0, I've removed this check. It was there to handle special cases of field, but this should be handled elsewhere now.